server/controllers/postUser.js

```diff
@@ -1,43 +1,45 @@
 const postUserQuery = require('../database/queries/addUsers');
-
-const joi = require('@hapi/joi');
-const validationObject = {
-  username: joi // username should be string and at least it contains 2 characters
-    .string()
-    .min(2)
-    .max(40),
-  email: joi // email should be string and valid and its tld should be .com
-    .string()
-    .email({
-      minDomainSegments: 2,
-      tlds: { allow: ['com'] },
-    })
-    .min(10),
-  password: joi // password must be string, at least have 6 characters contains capital letters, small letter and numbers
-    .string()
-    .alphanum()
-    .min(6)
-    .pattern(new RegExp('^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])')),
-};
-
-const schema = joi.object(validationObject);
+const checkEmail = require('../database/queries/ckeckEmail');
+const checkValidation = require('./validation');
 
 exports.postUser = (req, res) => {
-  const { error, value } = schema.validate(req.body);
-  const { username, email, password } = value;
-  if (error) {
-    res.status(400).json({ error: error.details[0].message });
-  } else {
-    try {
-      const userInfo = {
-        username,
-        email,
-        password,
-      };
-      postUserQuery(userInfo);
-      res.status(200).json({ message: 'user  created succefully' });
-    } catch (err) {
-      res.status(400).json({ message: 'There is no valid information' });
-    }
-  }
+  const { email } = req.body;
+  // check if email exists or not
+  checkEmail(email)
+    .then((result) => {
+      // if email doesn't exist create a new User
+      if (result.rows.length === 0) {
+        // check if email domain is gazaskygeeks.com
+        if (/@gazaskygeeks.com\s*$/.test(email)) {
+          const { error, value } = checkValidation(req.body);
+          // if there is an error in the validation send a message to client
+          if (error) {
+            res.status(400).json({ error: error.details[0].message });
+          } else {
+            // if there is no error in validation insert the user information in database
+            try {
+              const { username, email, password } = value;
+              const userInfo = {
+                username,
+                email,
+                password,
+              };
+              postUserQuery(userInfo);
+              res.status(200).json({ message: 'user  created succefully' });
+            } catch (err) {
+              res
+                .status(400)
+                .json({ message: 'There is no valid information' });
+            }
+          }
+        } else {
+          res
+            .status(400)
+            .json({ error: 'email domain must be gazaskygeeks.com' });
+        }
+      } else {
+        res.status(400).json({ message: 'email already exists!' });
+      }
+    })
+    .catch((err) => err);
```

Comment:
So your code is workable but definitely not maintainable nor is it readable.  I know you are short in time and so... still try your best to abstract it hence you are in **reactor week**.
I will give a hint or a little guide to do so... for example try to be consistent between using `async await` or `promise then`
Suppose your choice is `async await`, here is how a better code structure would be: 


```
try{
//check if email already exist if not throw an error with specific type/number/code
//validate data from client if not throw an error corresponding to the type of error
// insert the user if not successful throw an error with specific type/number/code
//res.send("succes...etc..)
}
catch(e){
 let error;
switch(e.code){
 case 1:
error="Email already exist");
break;
.....etc
}
res.state(400).json({message:error})
}
````

Reply:
you can helper function/s to help you get you to the level of abstraction mentioned above

Reply:
@ghassanmas I refactored the code using middlewares and helpers. Is it better now?